### ISSUE TITLE
Tomcat image not working on some nodes

### ISSUE THREAD

**Issue by MKlimuszka** [rancher/rancher#38170]
Issue description:

Trying to deploy a tomcat application on a downstream RKE cluster with Kubernetes 1.20.10.
They were previously using the SUSE CaaSP and the application worked fine.

Now, the application is failing on a few nodes in the cluster and it is working on others.

The nodes are the same ( kernel, OS, FS, patches, envs, locales )

CaaSP4.5 used cri.o as container runtime.
Their nodes are all using docker 20.10.6.
 
The application is a deployment with multiple replicas.

"On some nodes, the application fails to start because the libraries are started in the wrong order."

Q&A:

1/ How do you define the order of the libraries ?
The order is not explicity defined. It always be alphabetically.

2/ How do you detect the wrong lib loading order?
We get some “NoSuchMethodError” because the application use the wrong library and some methods could not exist.

3/ What is the messages indicating an issue with library loading?
“NoSuchMethodError”

4/ Is there a way to make it work in a single container runtime?
this is actually where the problem most probably come from, the container runtime is responsible for loading the FS and running the app.
You could use a tomcat image with a simple app that include two versions of the same library. If the app uses a method that exists only in one version of the library, you should be able to reproduce the problem easily.

 

Troubleshooting steps:
We have the Helm chart, the image and the Dockerfile used to build the image.
We can't start correctly the application because it depends on internal services (AS400 connection)  to start correctly.

I tried to start it from a container runtime ( Docker ) but the application relies on Kubernetes to start correctly.

 
Repro steps:
Cannot really repro the issue because of the image and internal required connections.

However, I found this bug on the Tomcat Bugzilla: https://bz.apache.org/bugzilla/show_bug.cgi?id=57129

It describes exactly the issue they are running into.

TLDR:
It seems like tomcat9 actually does NOT load libraries in alphabetical order.
It was alphabetical in Tomcat7, the libraries was sorted by Tomcat, but it changed since Tomcat8.

    https://bz.apache.org/bugzilla/show_bug.cgi?id=57129#c25
    WEB-INF/classes will be loaded before WEB-INF/lib/*.jar but there is no explicit guarantee of the ordering among the JAR files.

    https://bz.apache.org/bugzilla/show_bug.cgi?id=57129#c17
    Everything was deployed and working successfully in tests environments and 3 prod servers.
    However, the deployment was failing in the two prod servers. The error faced is 'java.lang.NoSuchMethodError'.
     

So now, everything converges in the data being loaded randomly.
 

Actual behaviour:
The application is failing to start on some nodes.

Expected behaviour:
It is expected to start and work on all the nodes the same way.

SURE-3617

**Comment by rohitsakala** [comment]
Closing since the relevant JIRA issue is closed.
